### ISSUE TITLE
Create async versions of INavigateAsync.RequestNavigate

### ISSUE THREAD

**Issue by wldevries** [PrismLibrary/Prism#118]
Using callbacks to check navigation request results is not really straight forward and feels too much like coding Javascript in C#. It would feel much more native to add awaitable versions. I created a gist of what the implementation could look like.

https://gist.github.com/frogger3d/b5820e10e715a3baa262

**Comment by briannoyes** [comment]
Another good one @frogger3d - would you mind submitting these as pull requests (along with your interaction request one)?

**Comment by YehudahA** [comment]
q: Navigation is a UI thread task. So, what asynchronous can give us?!

**Comment by brianlagunas** [comment]
Good question.

**Comment by briannoyes** [comment]
Same benefit an async dialog does - RequestNavigate does not complete until the current ViewModel allows it if it implements IConfirmNavigationRequest. During which time it can pop dialogs, interact with the user, etc. Having the calling code be able to say await RequestNavigate would just be a lot cleaner if the calling code does not have anything else sensible to do until navigation completes than having to provide a callback.

Always keep in mind that await Task does not mean "block thread", it means "pause this method's execution until the async thing is done".

**Comment by YehudahA** [comment]
Perhaps, but it seems a little strange and a little confusing to use async here.
On second thought, I agree that [for example] InteractionRequest.RaiseAsync is better than callback.

**Comment by likemike91** [comment]
Are there currently any plans to support async/await in the whole navigation process (INavigationAware, INavigateAsync, ...)?

**Comment by brianlagunas** [comment]
In Xamarin.Forms this already exists.  For WPF, this would be a major effort that is not a high priority.

**Comment by likemike91** [comment]
Okay I can imagine that it has a huge impact on the whole navigation system.

Are there any known issues with using INavigationAware respectively INavigateAsync with async void implementations?

**Comment by brianlagunas** [comment]
No known issues.

**Comment by wldevries** [comment]
There are differences in behavior:

When the `OnNavigatedTo` is non-async any error will be caught and cause a `NavigationFailed` event.
When the `OnNavigatedTo` is async any error will not be caught, the navigation will succeed and the Exception will be propagated to `Dispatcher.UnhandledException`.

Also if you use async void methods you cannot be sure that the actual navigation handling logic has finished when the `IRegionManager.RequestNavigate` has finished or when listening for the `Navigated` or `NavigationFalied` events. The reason for this is that async void methods complete immediately and are scheduled for deferred execution.

**Comment by dvorn** [comment]
In the perfect async world we would have
`public interface IConfirmNavigationRequest : INavigationAware`
`{`
`    Task<bool> ConfirmNavigationRequestAsync(NavigationContext navigationContext);`
`}`
and then
`public interface INavigateAsync`
`{`
`    Task<NavigationResult> RequestNavigateAsync(Uri target);`
`    Task<NavigationResult> RequestNavigateAsync(Uri target, NavigationParameters navigationParameters);`
`}`
with the corresponding changes in implementation of all classes that implement INavigateAsync.

This change is straightforward and can be easily done.  The major problem, however, is backward compatibility.

For classes that already implement IConfirmNavigationRequest we may leave it in place and give a new (Tasked) version of the interface a new name (e.g., IConfirmNavigationRequestAsync).  The implementation of navigation service will look for both interfaces and, if any of them is implemented, use it.

For existing programs that use old-style navigation requests we may create extension methods in the following style:

`public static async void RequestNavigate(this INavigateAsync navigator, Uri target,` `Action<NavigationResult> navigationCallback)`
`{`
`    NavigationResult result = await navigator(target);`
`    navigationCallback(result);`
`}`

However, there still be some compatibility issues for users that e.g. implemented their own IRegionNavigationService or created some mock objects for tests based on old interfaces.

@bartlannoeye @briannoyes @brianlagunas What do you think?

**Comment by brianlagunas** [comment]
> This change is straightforward and can be easily done.

This is not the case.  I looked into doing something like this and it was not a simple task, and was going to take forever, so I decided not to work on it as I had higher priorities.

This would definitely be great to have, but I just don't have the time to do it.

Having said that, after Prism for XF is ships as stable, I am going to be researching adding regions to XF and UWP, which means the creation of a new IRegionNavigationService that can be placed in the PCL and shared across all platforms.  Of course a lot of thought and planning would have to go into a feature like this across all platforms.

**Comment by dvorn** [comment]
@brianlagunas What problem do you see implementing this?  I see no one.  The only problem, as I said, is that with new interfaces it will not be backward compatible.

**Comment by TaviTruman** [comment]
I have to side with Brain w.r.t. augmenting the PRISM code base with this capability set (**Asynchronous Task, Eventing & Coordination**); it is not a simple change. Brain you may recall that we are using PRISM on our projects and there a lot of parallel-tasking going on the background while our WPF and UWP apps are running. It took redesign of Regions concept to get what we needed. It's not appropriate for me to go into the design and implementation details here but taking on this type of change was not a simple task at all.

**Comment by dvorn** [comment]
@TaviTruman What "redesign" are you talking about?  Navigation _was designed from the very beginning_ to be asynchronous, this is why it supplies a callback argument.  And async/await only makes it more explicit.  That callback is exactly a _continuation_ after the await.

**Comment by brianlagunas** [comment]
@dvorn Region navigation in WPF is not really Async, and never was.  It was faked by providing a call back that would execute when navigation completed.  But, the navigation blocked the UI and was not done on a separate thread or in an async manner.

**Comment by dvorn** [comment]
@brianlagunas I disagree.  Region navigation in Prism _IS_ really asynchronous.  It was not "faked" but used a callback technique since async/await was not available at the time.  Navigation _does not block_ the UI thread, RequestNavigate returns before navigation is completed without blocking the UI thread.

Use of async/await does not necessarily imply execution on a separate thread.

**Comment by dvorn** [comment]
It seems you need a PR.

**Comment by brianlagunas** [comment]
you cannot update ContentControl.Content without blocking the UI.  That's done on the UI thread.

I would love to a PR, though @wldevries made an initial attempt and it was closed because we haven't done a proper review of the necessary changes that need to be made in order to support this without breaking anyone.

**Comment by TaviTruman** [comment]
@dvorn we have written our own async navigation framework extending it beyond the navigation services that Brian wrote in support of nested regions. We needed a faster more lightweight memory model and we heavily extended the concept of regions. Our problem was not the async nature of the navigation implementation; we require more actions to fire in relationship to navigation eventing and ui/ux workflow. We also added a global KB repository to host messaging between viewmodels and all region management metadata. The point is that what I thought was an easy extension to navigation turnout to be more complex.

**Comment by TaviTruman** [comment]
@dvorn I did submit a PR but pull it back after talking more with Brian. We will submit our work once we have a few more demos written.

**Comment by dvorn** [comment]
@brianlagunas This is the implementation I have in mind: https://github.com/dvorn/Prism/commits/AsyncNavigation

Note that the work is not finished:

- Method comments not updated
- Tests needs to be updated
- Need to support both (old and new) variants of IConfirmNavigationRequest

**Comment by brianlagunas** [comment]
@dvorn at first glance it looks pretty good.  Things to keep in mind; we can't introduce breaking changes. We have to make sure the NavigationJOurnal works properly.  We have to make sure the navigation events fire when they are supposed to. We have to make sure we cover IRegion interface as well.  I am sure there are some other things, but that's it off the top of my head.

**Comment by dvorn** [comment]
Take a look at how RegionNavigationService implementation was simplified/streamlined using async/await: https://github.com/dvorn/Prism/blob/AsyncNavigation/Source/Wpf/Prism.Wpf/Regions/RegionNavigationService.cs#L134  This was not possible without await support in the compiler, and the authors of the original code were really clever guys to do it using callbacks technique.

**Comment by dvorn** [comment]
@brianlagunas 

>  Things to keep in mind; we can't introduce breaking changes.

I would like to be explicit about that.  What do we consider a breaking change?  E.g., in [release notes for Prism.Wpf 6.0](https://github.com/PrismLibrary/Prism/wiki/Release-Notes---6.0.0) the change to the IRegionManager interface is mentioned but not marked as a breaking change.

Ordinary user applications such as accompanying samples will continue to work.  However, if the user implemented some [extension to Prism](https://github.com/PrismLibrary/Prism/blob/master/Documentation/WPF/AppendixE1-ExtendingPrism.md) they may well implement their own version of IRegionManager, which is explicitly mentioned as a possibility in the document.  Then this extension will break since the interface has changed.

The same situation is here with async navigation requests: IRegionManager and INavigateAsync interfaces has changed.  Any user code that invokes methods on (old) versions of these interfaces will continue to work since the implementation of those old methods is also provided (as extension methods).  However, any direct user implementation of those interfaces will be broken.

**Comment by dvorn** [comment]
@brianlagunas  Gentlemen, I need your advice on how to organize my repositories and branches.

I already have a fork of Prism in which I have the AsyncNavigation branch where I commit my async changes.  This branch will eventually become a PR if it will be accepted.

Beside that I would like to provide to samples of the usage of the modified library, based on [View-Switching Navigation Sample](https://github.com/PrismLibrary/Prism-Samples-Wpf/tree/master/View-Switching%20Navigation_Desktop).  The first one will be un-modified sample linked to modified Prism library to demonstrate that it is not broken.  The second one will be a modified version which will demonstrate the use of the modified IConfirmNavigationRequest and some async/await-related simplifications.

My current plan is to create two more branches in my fork of Prism - OriginalSample and ModifiedSample.  In the former I will copy the navigation sample into the Sandbox and change references to local Prism projects.  The latter will also contain a copy of navigation sample with modified references, but in this branch I will modify the sample.

This layout will allow everyone to easily download the source code, compile and run the samples.

The drawback is that when and if the original navigation sample is modified, I will need to manually copy the changes.   Also, I would like to eventually create a PR from the modified sample (provided that the PR for Prism is accepted).  This will also be difficult to do.

But probably there is a better approach?

**Comment by brianlagunas** [comment]
@dvorn  I don't mind breaking custom implementations (adding to interfaces), we just can't break default implementations.

As far as samples go, the Sandbox folder is not meant to be actual samples, but rather just a playground for developing features.  They can be completely deleted at anytime.

If we can get the PR to a point where it is accepted, then you should fork the WF Samples repo, and your sample there.  Then when the NuGet packages are updated, switch it to use the public Nugets, and then submit the sample as a PR in the samples repo.

**Comment by dvorn** [comment]
@brianlagunas @briannoyes @bartlannoeye @wldevries See a short description of proposed changes in the description of PR #646

**Comment by sunefred** [comment]
@dvorn It is a shame this project died out in June. Did you guys decide to wait for a more general solution? From what I can deduct, the core of your solution is this line:

`bool canNavigate = await vetoingView.ConfirmNavigationRequestAsync(navigationContext);`

The rest of the changes basically just domino effects of using await on that line. By being selective of your await you elegantly tip-toe around the ContentControl.Content = ... that is buried somewhere in Region.Activate()

Now, assuming I have the gist of the suggested solution I would request that the entire INavigationAware interface is made async as well. It could be done totally analogous to your solution by having an INavigationAwareAsync interface .This is actually the route we are going in our project. The reason is that we do quite a but of view model initialization in OnNavigatedTo(). We could possibly spin up a background thread for this, but having the entire navigation chain be async is much more elegant.

**Comment by brianlagunas** [comment]
@sunefred It didn't die. I am only one person. I haven't had time to properly investigate this implementation or all of the possible repercussions of implementing it.  Think more of it as being in a coma. One day it will have a miraculous awakening :)

**Comment by dvorn** [comment]
@sunefred The INavigationAware interface cannot be async.  See discussion in #704

**Comment by kylewarren** [comment]
Is there anything wrong with calling your own methods? 

For instance ...
Public async void onNavigatingTo(NavigationParameters params)
{
 await OnNavigatingToTask(params)
}

Public async Task OnNavigatingToTask(NavigationParameters params)
{
 await doAsyncStuff();
}

**Comment by sunefred** [comment]
@brianlagunas It makes me sad then that cloning research is not further along then =)

@kylewarren This is a workable solution for some scenarios, but there are some limitations directly related to `async void`

1. Exceptions will not be repackaged and thrown up the hierarchy. This puts a requirement to handle the exception right there in OnNavigatedTo. This is doable depending on your application design, but some applications might rely on a "catch all" at the top to log and show simple dialogs etc.
2. Since `async void` is essentially a "fire and forget" function your navigation will no longer _respect the lifecycle_ that PRISM otherwise uses. This means that the navigation callback and the IRegionNavigationService navigated event could be called before your initialization code is finished.

[2] is the most sever IMO. The code late in your life-cycle could be made to wait for the your "fire and forget" code to finish, but then waiting has to be made `async`. It is all around just messy. If you instead implement INavigationAwareAsync the life-cycle is naturally respected.

@dvorn you make an excellent point that `async` implementation of this interface requires the user of such an interface to handle (or prevent) interactions to both the old and new views as the switch happens. For long running synchronous code the developer should already be aware that something extra has to be done to inform the user of the switch, so the leap to prevent interaction should not be so big. Most importantly though `async` will allow your "Loading..." message to be responsive and animating to and allow for better feedback to the user.

**Comment by kylewarren** [comment]
@sunefred what is the best route for handling async calls in the navigation methods? Just don't have them be async voids and don't await? Or force them to be synchronous? 

I can see how you could use IConfirmNavigation as a way to handle async calls that need to be done before onNavigatedFrom fires, but not others.

**Comment by brianlagunas** [comment]
> the developer should already be aware that something extra has to be done to inform the user of the switch

People have a hard enough time with the simplest features of Prism.  Implementing this async approach could be a big can of worms I'm not sure I want to open.  I'm trying to decide what the real benefit of this would be.  There is nothing wrong with async void, and in the case of INavigationAware it would make sense.  The INavigationAware methods can be considered events, but due to the nature of the extensibility they we made an interface instead.  Async void dosn;t really do anything except force you to do proper error handling.

Beside syntactical sugar, what is the real need for this feature?  What does not having this prevent your from doing?

**Comment by wldevries** [comment]
I completely agree @sunefred.

In addition to what you describe in point [1] the design also fails to meet the fail-fast principle. When your code fails you expect it to throw an exception right then and there. Instead you need to add a try/catch block or handle the NavigationFailed event. We've had an instance where half of the `OnNavigatedTo` on a viewmodel failed causing an extremely rare and hard to reproduce bug to pop up. Developers starting with Prism will be caught off-guard with this design. Swallowing exceptions is bad.

As a third point I'd like to add that when you call RequestNavigateTo you want to be able to get the NavigationResult then and there by awaiting the result. The addition of local functions makes it slightly easier to handle the callback in the current design, but it's far from nice experience to work with.

**Comment by brianlagunas** [comment]
> Instead you need to add a try/catch block or handle the NavigationFailed event

If you're not doing this in your production apps, you're doing it wrong.  Catch all is a horrible way to handle exceptions, and I would never allow that in my apps.  Is you are using a Tasked based approach, and an exception is thrown, you still have to handle it or it will be swallowed.  Nothing changes here.

**Comment by sunefred** [comment]
> Catch all is a horrible way ...

Yes yes, for sure, I don't think you will have disagreement there from either of us. It does seem useful to be able to do the following asynchronously

```
try
{
    _regionManager.RequestNavigate();
}
catch
{
    // Alert user of navigation failiure
}
```

This is a practical use case where `async `would be of direct use.

You did challenge us to give you the killer use case that cannot be accomplished without `async`. What we are looking for is a fully working set of lifetime hooks where the sequence and behavior is sequential and predictable. Consider the lifetime hooks Angular 2 provides for different stages of a Components creation and destruction. An essential quality of such a system is that the order in which they are executed should be strictly sequential, meaning hook #1 is finished _executing_ before #2 is called. You don't want the "Loaded" event before the "Initialize" event has finished executing.

You could argue that this is not what PRISM provides and then describing the event based view makes a lot of sense, and it that world I do not see a need for `async`.

**Comment by wldevries** [comment]
That is exactly the problem! If you don't add a try/catch block or handle the navigationfailed Prism will silently ignore any exceptions in your OnNavigatedTo, there is no other choice.

**Comment by brianlagunas** [comment]
@wldevries Exception handling is not a motivating factor for implementing this async feature.  If that is the basis for this feature, then it will not be implemented.

**Comment by dvorn** [comment]
Unfortunately Prism for WPF is dead.

**Comment by dansiegel** [comment]
@dvorn I would have to disagree. Prism has an active community. I would say that Prism for WPF is stable which is why you don't see as many changes.

**Comment by dvorn** [comment]
@dansiegel I would say that Prism for WPF fell the victim of its own success.  It has too large user base for the maintainers have a courage to change anything.

**Comment by dvorn** [comment]
Another example: user a asks for functionality on [SO](http://stackoverflow.com/questions/43691312/how-to-remove-single-view-from-prism-navigation-journal) and the solution exists for an year https://github.com/PrismLibrary/Prism/issues/30 and there is no movement in Prims.WPF

**Comment by brianlagunas** [comment]
@dvorn that is a knee-jerk comment that is unfounded an disrespectful.  You of all people should see the amount of work I put into this project every day.  Honestly, coming from you, it pisses me off.  Let's get a few things straight.  First, every single release of Prism.Wpf so far has had bug fixes and small updates.  Also, just because someone has a feature request or an idea doesn't mean it will be accepted.  Prism has a massive install base and is used in mission critical, and life-saving software.  Everything from finance, banking, infrastructure, medical, 911 software, and nuclear applications just to name a few.  You do not know the scope of Prism and don't talk to, or meet, with the customers in these industries.  I do!  For your information, no one else is asking for this feature (two doesn't count).  It is a very high risk feature that has no real demonstrated benefit, and needs to be properly and thoroughly tested before it will even be considered for implementation.  I haven't had time to do this.  The fact that I haven't closed it is showing good faith that the feature is still under consideration.  After that stupid-ass comment I am very motivated to just close it and move on.

Since you think Prism.Wpf is dead, I suggest you fork it, and create your own version that you support 24 hours a day 7 days a week. You can travel the world evangelizing your own framework and deal with ungrateful people that don't get what they want when they want it.  You can spend all your free time away from family and friends writing tooling and training material to help your users.

Either contribute and support the project the best you can, or don't.  That's up to you. What you won't do is make unfounded and ludicrous claims about a project that is constantly being worked on as "dead", just because one of your PR's aren't being merged quickly enough.  The choice is yours.

**Comment by sunefred** [comment]
@brianlagunas Hey man, being new to Prism this post was actually very informative. I had no idea of extent or impact of Prism in the industry and this sheds some light on that. I have not been around when it was part of Microsoft Patterns & Practices and have not gone through any versions prior to 6.0. It is easy to cook up a feature and get it working in a controlled environment, in a local project, but as the maintainer you always have to keep the big picture in mind and preserve backwards compability. This is an order of magnitude harder and your work here _is_ appreciated. Thank you.

Having said that, we are implementing features that could be useful for a framework like Prism but most likely would break backwards compability. Maybe there is room for a major (breaking) version down the line where such features could be considered? Other than async navigation, we have implemented a system for relative navigation and are working on scoped service location. (Lifetimescope in Autofac speak). I will elaborate on this in a separate topic if there is interest. Regardless, what is your view on breaking changes in major versions?

**Comment by brianlagunas** [comment]
@sunefred You hit the nail on the head, and thank you for your support. I think a conversation is always worth having.  One thing that I am very protective about is behavior changes.  I am not apposed to making breaking changes as long as they are obvious and can easily be found such as an API break.  Behavior changes are the most risky and challenging to introduce.  If the break can be found with a compile error, that is more likely to be considered than a change in logic that will be impossible to discover until the app is in production and being used.  Those are the dangerous and critical breaks.  There must also be a demonstrated benefit to Prism in order to add the feature.

I suggest you open your issue and start the discussion.  We can work through all your ideas and what problems we are trying to solve.

**Comment by dvorn** [comment]
@brianlagunas I did not meant to hurt you and I apologize if I sounded rude.  English is not my native language and to be understood I sometimes explain myself in straightforward and explicit words.  Of course I know how much you do to the Prism and respect that.

**Comment by brianlagunas** [comment]
@dvorn apology accepted. I understand that it's easy to get frustrated when things you are passionate about don't move fast enough.

As far as adding this feature goes, instead of trying to modify what's there and breaking existing apps, what about if we introduce an new "opt-in" region service called IRegionNavigationService which is async friendly?

**Comment by brianlagunas** [comment]
I am closing this as it is a duplicate of the more focused #657. We will continue any discussion there.
